File: Src/Runtime/Strings/Format.cpp
/*
 * #-----------------------------------------------------# *
 * #                                                     # *
 * #                           Typed                     # *
 * #                   A text formatting DSL             # *
 * #                                                     # *
 * #-----------------------------------------------------# *
 * #                                                     # *
 * #         Created by Rodrigo R. & Contributors        # *
 * #         Released under the Apache License 2.0       # *
 * #            Check LICENSE.MD for more info           # *
 * #                                                     # *
 * #-----------------------------------------------------# *
*/

//
// Created by Rodrigo on 5/27/26.
//

#include "Format.h"
#include "ADT/List/Object.h"
#include "Support/Format/Format.h"

#include "Support/Runtime/AccessString.h"

using namespace Typed;
using namespace Typed::Runtime;
using namespace Typed::Runtime::Strings;

ADT::Runtime::Object Strings::Format(
    ADT::List::Object &args,
    ADT::Lang::AST *trace
)
{
    auto &obj = args[0];
    auto str = Support::Runtime::AccessString(obj);
    Celery::Str::String result;

    Support::Format::Format(
        str,
        args,
        result,
        trace->line,
        trace->column
    );

    return {
        ADT::Runtime::ObjectType::OwnedString,
        std::move(result)
    };
}

File: src/cli/command/format.rs
/*
 * #-----------------------------------------------------# *
 * #                                                     # *
 * #                           Typed                     # *
 * #                   A text formatting DSL             # *
 * #                                                     # *
 * #-----------------------------------------------------# *
 * #                                                     # *
 * #         Created by Rodrigo R. & Contributors        # *
 * #         Released under the Apache License 2.0       # *
 * #            Check LICENSE.MD for more info           # *
 * #                                                     # *
 * #-----------------------------------------------------# *
*/
use crate::cli::FormatCommand;
use crate::core::formatter::format;
use crate::support::failable::{catch_non_traceable};
use crate::support::file;

pub fn handle(cmd: &FormatCommand) {
    let contents = file::read(cmd.file.as_str());
    let contents = catch_non_traceable(&contents);
    let ast = file::parse(&contents, &cmd.input_syntax);

    let fmt = format(&ast, &cmd.output_syntax);
    let res = file::write(cmd.output.as_str(), &fmt);

    if res.is_err() {
        panic!("{}", res.err().unwrap());
    }
}